pallets/loans/src/entities/loans.rs

```diff
@@ -258,17 +261,31 @@ impl<T: Config> ActiveLoan<T> {
 
 					let now_date = NaiveDateTime::from_timestamp_opt(now as i64, 0)
 						.ok_or(DispatchError::Other("Invalid now date"))?;
-					let cashflows_in_past = cashflows
-						.iter()
-						.filter(|(d, _)| d.and_hms_opt(0, 0, 0).unwrap() > now_date);
-
-					// TODO: this is incorrect. It should count the # of interest payments
-					// that are overdue based on self.total_repaid and the grace period
-					let expected_total_repaid: T::Balance = cashflows_in_past
-						.map(|(_, a)| a)
-						.try_fold(T::Balance::zero(), |a, b| a.ensure_add(*b))?;
-
-					Ok(false)
+					let cashflows_in_past = cashflows.iter().filter(|(d, _)| {
+						d.and_hms_opt(0, 0, 0).unwrap()
+							> now_date + Duration::seconds(*grace_period as i64)
+					});
+
+					let (cashflows_paid_count, _): (u32, _) =
+						cashflows_in_past.clone().map(|(_, a)| a).fold(
+							(0u32, self.total_repaid),
+							|(cashflows_paid_count, remaining_repaid), cashflow| {
+								if remaining_repaid > *cashflow {
+									return (cashflows_paid_count, T::Balance::zero());
+								} else {
+									return (
+										cashflows_paid_count + 1,
+										remaining_repaid
+											.ensure_sub(*cashflow)
+											.expect("TODO: this is unsafe"),
```

Comment:
You can use `try_fold()` instead and extract these arithmetic errors from the iteration.